### PR TITLE
Feature/rdb block store

### PR DESCRIPTION
**Pull request by iceseer**
### Description of the Change
Adds block store based on ```RocksDB```.

### CODE REVIEW

File: irohad/ametsuchi/impl/rocksdb_block_storage.cpp [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        uint64_t count = 0ull;
+        if (auto result =
+                forBlocksTotalCount<kDbOperation::kGet, kDbEntry::kMustExist>(
+                    common);
+            expected::hasValue(result))
+          count = *result.assumeValue();
+
+        common.encode(count + 1ull);
+        if (auto result =
+                forBlocksTotalCount<kDbOperation::kPut, kDbEntry::kMustExist>(
```
Comment (kamilsa) [inline irohad/ametsuchi/impl/rocksdb_block_storage.cpp:60]: Does it make sense to store block itself and block's count atomically?

**Review by kamilsa — APPROVED** [review]

File: irohad/ametsuchi/impl/rocksdb_block_storage.cpp (hunk 1 above)
Comment (iceseer) [reply irohad/ametsuchi/impl/rocksdb_block_storage.cpp:60]: It's atomic. RocksdbCommon ctor make lock.

File: irohad/ametsuchi/CMakeLists.txt [hunk 2]
```diff
@@ -201,6 +214,7 @@ target_link_libraries(ametsuchi
     rdb_connection_init
     ametsuchi_rocksdb
     flat_file_storage
+    rocksdb_block_storage
```
Comment (kuvaldini) [inline irohad/ametsuchi/CMakeLists.txt:217]: why flat file storege depends on rocksdb. It does not.

File: irohad/ametsuchi/impl/rocksdb_block_storage.cpp [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    std::shared_ptr<shared_model::interface::BlockJsonConverter> json_converter,
+    logger::LoggerPtr log)
+    : db_context_(std::move(db_context)),
+      json_converter_(std::move(json_converter)),
+      log_(std::move(log)) {}
+
+bool RocksDbBlockStorage::insert(
+    std::shared_ptr<const shared_model::interface::Block> block) {
+  return json_converter_->serialize(*block).match(
+      [&](const auto &block_json) {
```
Comment (kuvaldini) [inline irohad/ametsuchi/impl/rocksdb_block_storage.cpp:26]: could you please remake this spagetti. name this lambda as `handle_serialized_block()`, and the second one `handle_error()`

File: irohad/ametsuchi/impl/rocksdb_block_storage.cpp (hunk 1 above)
Comment (kuvaldini) [reply irohad/ametsuchi/impl/rocksdb_block_storage.cpp:60]: <kDbOperation::kPut, kDbEntry::kMustExist> is not obvious to understand.

Should the record exist before operation or after it. I expect it kMustExist before the operation, so must be rewritten. Please provide more obvious notation.

Comment (iceseer) [reply irohad/ametsuchi/impl/rocksdb_block_storage.cpp:60]: To make check before operation execution is a logic case. I asked you to add issue with extended ```kDbEntry``` checks for ```kPut``` and ```kDel``` operations.

File: irohad/ametsuchi/CMakeLists.txt (hunk 2 above)
Comment (iceseer) [reply irohad/ametsuchi/CMakeLists.txt:217]: ???
This is dependencies for ```ametsuchi```.

Comment (kuvaldini) [reply irohad/ametsuchi/CMakeLists.txt:217]: I thought ametsuchi_rocksdb should depend on rocksdb_block_storage

Comment (iceseer) [reply irohad/ametsuchi/CMakeLists.txt:217]: ```ametsuchi``` is not a flat file storage.

File: irohad/ametsuchi/impl/rocksdb_block_storage.cpp [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+              result.assumeError().description);
+          return false;
+        }
+
+        common.valueBuffer() = block_json.value;
+        if (auto result = forBlock<kDbOperation::kPut>(common, block->height());
+            expected::hasError(result)) {
+          log_->error("Error while block {} storing. Code: {}. Description: {}",
+                      block->height(),
+                      result.assumeError().code,
```
[the remark this answers is not in the document]
Comment (iceseer) [reply irohad/ametsuchi/impl/rocksdb_block_storage.cpp]: All of them means that something wrong with database or with logic.

File: irohad/ametsuchi/impl/rocksdb_block_storage.cpp [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    return *result.assumeValue();
+  return 0ull;
+}
+
+void RocksDbBlockStorage::reload() {}
+
+void RocksDbBlockStorage::clear() {
+  RocksDbCommon common(db_context_);
+
+  if (auto status = common.filterDelete(fmtstrings::kPathWsv); !status.ok())
```
[the remark this answers is not in the document]
Comment (iceseer) [reply irohad/ametsuchi/impl/rocksdb_block_storage.cpp:113]: I don't want to allow to clear block storage without cleaning wsv.

File: irohad/ametsuchi/impl/rocksdb_block_storage.cpp [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+iroha::expected::Result<void, std::string> RocksDbBlockStorage::forEach(
+    iroha::ametsuchi::BlockStorage::FunctionType function) const {
+  uint64_t const blocks_count = size();
+  for (uint64_t ix = 1; ix <= blocks_count; ++ix) {
+    auto maybe_block = fetch(ix);
+    if (maybe_block) {
+      auto maybe_error = function(std::move(maybe_block).value());
+      if (iroha::expected::hasError(maybe_error)) {
+        return maybe_error.assumeError();
```
[the remark this answers is not in the document]
Comment (kuvaldini) [reply irohad/ametsuchi/impl/rocksdb_block_storage.cpp:128]: assumeError() also checks?

**Review by kuvaldini — APPROVED** [review]

File: irohad/ametsuchi/impl/rocksdb_block_storage.cpp (hunk 1 above)
Comment (kuvaldini) [reply irohad/ametsuchi/impl/rocksdb_block_storage.cpp]: #1330

**Pull request merged**